apos.1
.\" Manpage for nuseradd.
.\" Contact cdvivek@nixcraft.net.in to correct errors or typos.

.TH apos 1 "11 Jan 2020" "v2.0.1" "apos man page"

.SH NAME
apos -- create an empty project for AVR microcontrollers.

.SH SYNOPSIS
apos [OPTIONS] <project_name>

.SH DESCRIPTION
.B apos
creates an AVR microcontroller project template for Unix/Linux systems based on a Makefile using 
.BR avr-gnu " toolchain (avr-g++)", 
.BR avrdude " and" 
.BR make " to build aund upload the .hex file into microcontroller."

.PP
This project is inspired in the 
.BR avr-proj " program That is include in the CrossPack AVR software package for MacOS."
It offers a free alternative with improvements and extra options, which works with Linux and MacOS operating systems in a simple but very useful program.

.PP 
apos creates a folder with the given project name. Inside the folder the
.BR main.c ", " Makefile " and " README.md " files are created, along with the lib folder."

.PP
If you need to change the microcontroller frequency, model, programmer, fuses or another configuration,  edit the lines from 6 to 14 of the project Makefile:

  ...
  5
  6 PROJECT_NAME  = my_project
  7 DEVICE        = atmega128
  8 CLOCK         = 16000000
  9 FUSES         = -U hfuse:w:0xD9:m -U lfuse:w:0xFF:m #-U efuse:w:0xFF:m
 10 AVRDUDE_PROG  = -c usbtiny -P usb
 11 ABOS_PORT     = /dev/tty.usbserial-1A12430
 12 ABOS_BAUDRATE = 38400
 13 AVRDUDE_OPS   = -B 0.5
 14 AVRDUDE       = avrdude -p $(DEVICE) $(AVRDUDE_PROG) $(AVRDUDE_OPS)
 15
 ...

.PP
Adding you own hardware drivers and libraries is very simple, only put your header and source files in the 'lib' directory or another directory.

The final step is uncomment and copy the lines 15 and 19 of the 'Makefile' to add your source files and include paths:
 
 ...
 15
 16 OBJECT_FILES = main.o
 17 #OBJECT_FILES += ./src/mySource.o
 18
 19 INCLUDEPATHS =  -I .
 20 INCLUDEPATHS =  -I ./lib
 21 #INCLUDEPATHS += -I ./myFolder
 22
 ...

.SH OPTIONS

-b          Creates a blink project template

-c          Defines the avrdude programmer (e.g usbasp, usbtiny, dragon_isp, etc.)

-e          Defines the Extended Fuse  (2 digit hex format)

-f          Define the CPU Frequency in Hz (default 16000000)

-g          Initialize a GIT repository in the project folder

-h          Defines the High Fuse  (2 digit hex format)

-l          Defines the Low Fuse (2 digit hex format)

-m          Define the Microcontroller (default: atmega128)

-p          Define the ABOS Loader serial port (default: /dev/ttyUSB0)

-s          Define the ABOS Loader baudrate (default: 38400)

-v          Shows the apos version

--help      Shows apos help

.SH MAKE TARGETS
 
make               Compile the source code.

make all           Same as make.

make clean         Clean all the compiled output files.

make reset         Reset the AVR microcontroller using avrdude.

make flash         Upload the generated .hex file into AVR microcontroller using avrdude.

make read_flash    Reads the AVR microcontroller flash memory and save its content in "*.flash.bin" file.

make fuses         Program the specified fuses into the AVR microcontroller.

make install       Program the fuses bits, flash and eeprom memories into the AVR microcontroller.

make eeprom        Program the eeprom memori with the .eep generated file.

make read_eeprom   Reads the AVR microcontroller eeprom memory and save its content in "*.eeprom.bin file.

make erase         Erase the flash memory in the AVR microcontroller using avrdude.

make disasm        Disassemble the .elf generated file.

make abos          Load the .hex file with the
.BR abosldr " command (if the ABOS bootloader is installed in the AVR)"

.SH EXAMPLES
  Empty project:              apos example
  With git repository:        apos -g example
  Using usbtiny programmer:   apos -c usbtiny example
  Create a Blink project:     apos -b blink
  Define MCU:                 apos -m atmega32 example
  Define CPU frequency:       apos -m atmega32 -f 16000000 example
  Define programmer:          apos -m atmega32 -f 16000000 -p usbtiny example
  Define fuses:               apos -m atmega32 -f 16000000 -p usbtiny -e ff -l de -h D9 example
  Define ABOS params:         apos -m atmega32 -f 16000000 -p /dev/ttyACM0 -s 38400 exampleq

.SH BUGS
No known bugs.

.SH AUTHOR
Alfredo Orozco (alfredoopa@gmail.com)

.SH SEE ALSO
abosldr(1), avrdude(1), avr-gcc(1)

CHANGELOG.md
# APOS Changelog

## [ 2.0.1 ] - APR 2020

#### Updated
  - Update the installation path for OSx.

#### Update
  - Change the help option, now is only --help. -? and ? cause problems in szh terminal.

## [ 2.0.0 ] - JAN 2020

#### Updated
  - Update examples in help screen.
  - Minor updates to default blink and makefile templates.

#### Changed
  - Change the name of command options to be only one letter.
     - avrdude programmer option "-p" now is "-c"
     - abos loader port option "-ap" now is "-p"
     - abos loader baudrate option "-ab" now is "-s"
     - extended fuse option "-ef" now is "-e"
     - high fuse option "-hf" now is "-h"
     - low fuse option "-lf" now is "-l"

#### Added
- Add default .gitignore file to proyectfolder with -g option.
- Add man page (1), now shows apos manual with *man apos*

## [ 1.8.0 ] - 2018-11-10

#### Added
- Add ABOS support to project creation
  - -ap: Specify the Serial Port for the ABOS Loader
  - -ab: Specify the Serial Port Baudrate for the ABOS Loader


## [ 1.7.0 ] - 2018-4-14

#### Added
- Add AVR eeprom code generation and programation.
- Add new Makefile targets
  - eeprom
  - read_eeprom
  - read_flash

#### Changed
* Changes the APOS license, now is GPL


## [ 1.6.0 ] - 2018-3-26

#### Updated
- Update APOS Help


## [ 1.5.0 ] - 2017-27-10

#### Updated
- Rename the project to APOS (AVR Project Open Source)


## [ 1.4.0 ] - 2017-27-10

#### Added
- Add git initialization to project creation
- Add blink template option
- Add options to configure the AVR model, frequency, avrdude programmer and fuses


## [ 1.3.1 ] - 2017-30-8

#### Added
- Add erase target to make file


## [ 1.3.0 ] - 2017-27-7

#### Added
- Add README.md

#### Updated
- Update the project structure to support basic templates


## [ 1.2.1 ] - 2017-8-7

#### Changed
- Change the default avrdude programmer to usbasp


## [ 1.2.0 ] - 2017-4-7

#### Added
- Add git initialization to project creation
- Add blink template option
- Add options to configure the AVR model, frequency, avrdude programmer and fuses


## [ 1.1.0 ] - 2017-22-6

#### Added
- Add include and src folders to better project organization

## [ 1.0.0 ] - 2017-21-5
First version of avr-proj tool for Linux

apos.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <string.h>
#include <time.h>
#include <ctype.h>

#include "templates.h"

#define RED				"\x1B[31m"
#define BOLD_RED		"\x1B[1;31m"
#define GREEN			"\x1B[32m"
#define BOLD_GREEN		"\x1B[1;32m"
#define YELLOW			"\x1B[33m"
#define BOLD_YELLOW		"\x1B[1;33m"
#define BLUE			"\x1B[34m"
#define BOLD_BLUE		"\x1B[1;34m"
#define MAGENTA			"\x1B[35m"
#define BOLD_MAGENTA	"\x1B[1;35m"
#define CYAN			"\x1B[36m"
#define BOLD_CYAN		"\x1B[1;36m"
#define WIHTE			"\x1B[37m"
#define BOLD_WIHTE		"\x1B[1;37m"
#define RESET			"\x1B[0m"
#define BOLD_RESET		"\x1B[1;0m"

#define VERSION  			"1.6.3"
#define AUTHOR  			"Alfredo Orozco de la Paz"
#define AUTHOR_EMAIL		"alfredoopa@gmail.com"
#define LAST_COMPILATION	"April 1 2018"

typedef struct {
	char blink_template;
	char create_git;
	char version;
	char help;
	char f_cpu[20];
	char low_f[20];
	char high_f[20];
	char extended_f[20];
	char device[50];
	char programmer[50];
	char project_name[1024];
}options_t;

char txt[2048];

void show_version(){

	printf( BOLD_MAGENTA "      ___           ___           ___           ___     									\n");
	printf("     /\\  \\         /\\  \\         /\\  \\         /\\  \\    						\n");
	printf("    /::\\  \\       /::\\  \\       /::\\  \\       /::\\  \\   						\n");
	printf("   /:/\\:\\  \\     /:/\\:\\  \\     /:/\\:\\  \\     /:/\\ \\  \\  				\n");
	printf(BOLD_CYAN "  /::\\~\\:\\  \\   /::\\~\\:\\  \\   /:/  \\:\\  \\   _\\:\\~\\ \\  \\ 		\n");
	printf(" /:/\\:\\ \\:\\__\\ /:/\\:\\ \\:\\__\\ /:/__/ \\:\\__\\ /\\ \\:\\ \\ \\__\\	\n");
	printf(" \\/__\\:\\/:/  / \\/__\\:\\/:/  / \\:\\  \\ /:/  / \\:\\ \\:\\ \\/__/			\n");
	printf("      \\::/  /       \\::/  /   \\:\\  /:/  /   \\:\\ \\:\\__\\  						\n");
	printf(BOLD_YELLOW"      /:/  /         \\/__/     \\:\\/:/  /     \\:\\/:/  /  								\n");
	printf("     /:/  /                     \\::/  /       \\::/  /   \n");
	printf("     \\/__/                       \\/__/         \\/__/    \n\n");


	printf(RESET GREEN " APOS" RESET " (AVR Project Open Source)"BOLD_CYAN" v%s\n" BOLD_RESET" Last compilation: " BOLD_CYAN "%s\n\n", VERSION, LAST_COMPILATION);
	printf(" Creates an AVR project template based on a makefile.\n Uses the avr-gcc toolchain and avrdude.\n\n");
}

void show_help() {
	show_version();
	printf(BOLD_RESET "   Usage:" BOLD_RED " apos " RESET "[" BOLD_YELLOW "OPTIONS" RESET "] <" BOLD_CYAN "Project_Name" RESET ">\n\n");
	printf(BOLD_YELLOW"   -?"RESET" or"BOLD_YELLOW" ? "RESET"option Shows help\n\n");

	printf(BOLD_YELLOW "--------------------------------------------\n\n" RESET);
	printf(BOLD_YELLOW "   OPTIONS:\n\n" RESET);
	printf(YELLOW "     -b" RESET "    Create a blink project template\n");
	printf(YELLOW "     -ef" RESET "   Define the Extended Fuse  (2 digit hex format)\n");
	printf(YELLOW "     -f" RESET "    Define the CPU Frequency (in Hz)\n");
	printf(YELLOW "     -g" RESET "    Initialize a GIT repository in the project folder\n");
	printf(YELLOW "     -hf" RESET "   Define the High Fuse  (2 digit hex format)\n");
	printf(YELLOW "     -lf" RESET "   Define the Low Fuse (2 digit hex format)\n");
	printf(YELLOW "     -m" RESET "    Define the Microcontroller (e.g atmega328p)\n");
	printf(YELLOW "     -p" RESET "    Define the avrdude programmer (e.g usbasp, usbtiny, dragon_isp, etc.)\n");
	printf(YELLOW "     -v" RESET "    Show the " RED "apos " RESET "version\n");
	printf(YELLOW "     -?,? " RESET " Show this help\n\n");
	
	printf(BOLD_CYAN "--------------------------------------------\n\n" RESET);
	printf(BOLD_CYAN "   Makefile Options:\n\n" RESET);
	printf(CYAN "     all"  	 RESET "        Compile all the source file and generates the .hex file\n");
	printf(CYAN "     clean"   RESET "      Clean the project\n");
	printf(CYAN "     erase"   RESET "      Erase the flash memory into the AVR\n");
	printf(CYAN "     fuses"   RESET "      Program the low, high and extended fuses\n");
	printf(CYAN "     install" RESET "    Program the fuses and flash memory into the AVR\n");
	printf(CYAN "     reset"   RESET "      Resets the AVR microcontroller\n");
	printf(CYAN "     disasm"  RESET "     Disassemble the .hex file and generate the .asm file\n\n");
	
	printf(BOLD_MAGENTA "--------------------------------------------\n\n" RESET);
	printf(BOLD_MAGENTA "   Examples: create a blink project\n\n" RESET);
	printf("     Empty project:            " RED "apos " CYAN "blink\n" RESET);
	printf("     Create a Blink project:   " RED "apos " YELLOW "-b " CYAN "blink\n" RESET);
	printf("     Define MCU:               " RED "apos " YELLOW "-m " RESET "atmega32 " CYAN "blink\n" RESET);
	printf("     Define CPU frequency:     " RED "apos " YELLOW "-m " RESET "atmega32 " YELLOW "-f " RESET "16000000 " CYAN "blink\n" RESET);
	printf("     Define programmer:        " RED "apos " YELLOW "-m " RESET "atmega32 " YELLOW "-f " RESET "16000000 " YELLOW "-p " RESET "usbtiny " CYAN "blink\n" RESET);
	printf("     Define fuses:             " RED "apos " YELLOW "-m " RESET "atmega32 " YELLOW "-f " RESET "16000000 " YELLOW "-p " RESET "usbtiny " YELLOW "-e "RESET"ff "YELLOW"-l "RESET"de "YELLOW"-h "RESET"D9 "CYAN"blink\n" RESET);
	printf("                \n");
	printf(RESET" Created by " BOLD_CYAN "%s" MAGENTA " <" YELLOW "%s" MAGENTA ">\n\n" RESET, AUTHOR, AUTHOR_EMAIL);
	
}

int parse_args(int argc, char const *argv[], options_t *options_out){

	int ix;

	memset(options_out, 0, sizeof(options_t));
	// Set the default options
	strcpy(options_out->device, "atmega328p");
	strcpy(options_out->f_cpu, "16000000");
	strcpy(options_out->programmer, "usbtiny");
	strcpy(options_out->high_f, "D9");
	strcpy(options_out->low_f, "DE");
	strcpy(options_out->extended_f, "FF");

	for(int i = 1; i< argc; i++)
	{
		ix=0;
		char const *option = argv[i];
		if (*option == '?')
		{
			options_out->help = 1;
			return 1;
		}
		if(option[0] == '-')
		{
			if(!option[1])
			{
				printf(RED "apos" RESET ": [" YELLOW "error" RESET "] invalid option " CYAN "%s\n\n" RESET, argv[i]);
				return 0;
			}
			switch(option[1]){
				case 'm': 
					if(i+1 < argc && argv[i+1][0] != '-') 
						strcpy(options_out->device, argv[i+1]);
					else
						strcpy(options_out->device, "missing");
				break;
				case 'l': 
					if(!option[2] || option[2] != 'f')
					{
						printf(RED "apos" RESET ": [" YELLOW "error" RESET "] invalid option " CYAN "%s\n\n" RESET, argv[i]);
						return 0;
					}
					if(i+1 < argc && argv[i+1][0] != '-') 
					{
						strcpy(options_out->low_f, argv[i+1]);
						while(options_out->low_f[ix])
						{
							options_out->low_f[ix] = toupper(options_out->low_f[ix]);							
							if(!isxdigit(options_out->low_f[ix]))
							{
								printf(RED "apos" RESET ": [" YELLOW "error" RESET "] invalid hex value " CYAN "\"%c\"" RESET " for low fuses\n\n", options_out->low_f[ix]);
								return 0;
							}
							ix++;
						}
						if(strlen(options_out->low_f) >2)
						{
							printf(RED "apos" RESET ": [" YELLOW "error" RESET "] invalid hex value " CYAN "\"%s\"" RESET " for low fuses, needs 2 digits\n\n" RESET, options_out->low_f);
							return 0;
						}
					}
					else
						strcpy(options_out->low_f, "missing");
				break;
				case 'h':
					if(!option[2] || option[2] != 'f')
					{
						printf(RED "apos" RESET ": [" YELLOW "error" RESET "] invalid option " CYAN "%s\n\n" RESET, argv[i]);
						return 0;
					} 
					if(i+1 < argc && argv[i+1][0] != '-') 
					{
						strcpy(options_out->high_f, argv[i+1]);
						while(options_out->high_f[ix])
						{
							options_out->high_f[ix] = toupper(options_out->high_f[ix]);
							if(!isxdigit(options_out->high_f[ix]))
							{
								printf(RED "apos" RESET ": [" YELLOW "error" RESET "] invalid hex value " CYAN "\"%c\"" RESET " for high fuses\n\n", options_out->high_f[ix]);
								return 0;
							}
							ix++;
						}
						if(strlen(options_out->high_f) >2)
						{
							printf(RED "apos" RESET ": [" YELLOW "error" RESET "] invalid hex value " CYAN "\"%s\"" RESET " for high fuses, needs 2 digits\n\n" RESET, options_out->high_f);
							return 0;
						}
					}
					else
						strcpy(options_out->high_f, "missing");
				break;
				case 'e':
					if(!option[2] || option[2] != 'f')
					{
						printf(RED "apos" RESET ": [" YELLOW "error" RESET "] invalid option " CYAN "%s\n\n" RESET, argv[i]);
						return 0;
					} 
					if(i+1 < argc && argv[i+1][0] != '-') 
					{
						strcpy(options_out->extended_f, argv[i+1]);
						while(options_out->extended_f[ix])
						{
							options_out->extended_f[ix] = toupper(options_out->extended_f[ix]);							
							if(!isxdigit(options_out->extended_f[ix]))
							{
								printf(RED "apos" RESET ": [" YELLOW "error" RESET "] invalid hex value " CYAN "\"%c\"" RESET " for extended fuses\n\n", options_out->extended_f[ix]);
								return 0;
							}
							ix++;
						}
						if(strlen(options_out->extended_f) >2)
						{
							printf(RED "apos" RESET ": [" YELLOW "error" RESET "] invalid hex value " CYAN "\"%s\"" RESET " for extended fuses, needs 2 digits\n\n" RESET, options_out->extended_f);
							return 0;
						}
					}
					else
						strcpy(options_out->extended_f, "missing");
				break;
				case 'f': 
					if(i+1 < argc && argv[i+1][0] != '-') 
						strcpy(options_out->f_cpu, argv[i+1]);
					else
						strcpy(options_out->f_cpu, "missing");
				break;
				case 'p': 
					if(i+1 < argc && argv[i+1][0] != '-') 
						strcpy(options_out->programmer, argv[i+1]);
					else
						strcpy(options_out->programmer, "missing");
				break;
				case 'v': options_out->version = 1; break;
				case '?': options_out->help = 1; break;
				case 'b': options_out->blink_template = 1; break;
				case 'g': options_out->create_git = 1; break;
				default: printf(RED "apos" RESET ": [" YELLOW "error" RESET "] invalid option " CYAN "%s\n\n" RESET, argv[i]);
				return 0;
				
			}
		}
	}

	if(argv[argc-2][0] == '-' && (argv[argc-2][1] == 'v' || argv[argc-2][1] == '?' || argv[argc-2][1] == 'b' || argv[argc-2][1] == 'g')){
		strcpy(options_out->project_name, argv[argc-1]);
	}
	else if(argv[argc-2][0] != '-' && argv[argc-1][0] != '-')
	{
			strcpy(options_out->project_name, argv[argc-1]);
	}
	else{
		if(!options_out->version && !options_out->help)
		{
			printf(RED "apos" RESET ": [" YELLOW "error" RESET "] missing project name\n\n" RESET);
			return 0;
		}
	}
	return 1;
}

void create_makefile(options_t *options) 
{
	FILE *f;
	// Creates the "Makefile" whit the content.
	strcpy(txt, options->project_name);
	strcat(txt, "/Makefile");
	f = fopen(txt, "w");
	
	fwrite(makefile_header,sizeof(char), strlen(makefile_header), f);
	fprintf(f, "\nPROJECT_NAME = %s\nDEVICE       = %s\nCLOCK        = %s\nPROGRAMMER   = -c %s -P usb\nFUSES        = -U hfuse:w:0x%s:m -U lfuse:w:0x%s:m #-U efuse:w:0x%s:m",
	options->project_name, options->device, options->f_cpu, options->programmer, options->high_f, options->low_f, options->extended_f);
	fwrite(makefile_body,sizeof(char), strlen(makefile_body), f);
	
	fclose(f);
}

void create_mainfile(options_t *options)
{
	FILE *f;

	time_t t = time(NULL);
	struct tm tm = *localtime(&t);
	
	strcpy(txt, options->project_name);
	strcat(txt, "/main.c");
	f = fopen(txt, "w");
	if(options->blink_template)
	{
		fprintf(f,"/******************************************************\n * main.c\n *\n * Program Name:  Blink\n *         Date:  %04d-%02d-%02d", tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday );
		fwrite(blink_program, sizeof(char), strlen(blink_program), f);
	}
	else
	{
		fprintf(f,"/******************************************************\n * main.c\n *\n * Program Name:  %s\n *         Date:  %04d-%02d-%02d", options->project_name, tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday );
		fwrite(empty_template_body, sizeof(char), strlen(empty_template_body), f);
	}
	fclose(f);
}

void create_gitrepo(options_t *options)
{
	sprintf(txt, "cd %s && git init && git add * && git commit -m \"Initial commit for APOS AVR project ~%s~\"",options->project_name, options->project_name);
	system(txt);
}

void create_readme(options_t *options) 
{
		FILE *f;
		// Creates RREADME.md
		strcpy(txt, options->project_name);
		strcat(txt, "/README.md");
		f = fopen(txt, "w");
		fprintf(f, "\n# %s \n\n### Project created with APOS (AVR Project Open Source)\n\nTo edit the projecto settings, open the Makefile and edit the firs lines of the file.\n\n", options->project_name);
		fclose(f);
}

void create_folder_project(options_t *options)
{
	FILE *f;
	// Replace the spaces in the project name by "_", to avoid compilation problems
	for (int i = 0; i < strlen(options->project_name); ++i)
	{
		if(options->project_name[i]==' ')
			options->project_name[i]= '_';
	}

	if(f=fopen(options->project_name, "r"))
	{
		printf(BOLD_RED "\napos" RESET ": [" YELLOW "error" RESET "] the project " BOLD_YELLOW "\"%s\"" RESET " already exist!\n\n", options->project_name);
		
		fclose(f);
		exit(1);
	}
	// Creates the project folder with the name passed in the program arguments
	mkdir(options->project_name, 0700);
	// Creates the "include" folder
	strcpy(txt, options->project_name);
	strcat(txt, "/include");
	mkdir(txt, 0700);
	// Creates the "src" folder
	strcpy(txt, options->project_name);
	strcat(txt, "/src");
	mkdir(txt, 0700);
}

void create_project(options_t *options) 
{
	create_folder_project(options);
	create_makefile(options);
	create_mainfile(options);
	create_readme(options);
	if(options->create_git)
		create_gitrepo(options);

}

int main(int argc, char const *argv[])
{
	int parse_result;
	options_t options;

	// If not pass arguments to program show the help
	if(argc < 2) {
		show_version();
		return 0;
	}

	parse_result = parse_args(argc, argv, &options);

	if(options.version)
	{
		show_version();
		return 0;
	}
	if(options.help)
	{
		show_help();
		return 0;
	}
	if(!parse_result){
		printf("Write  apos " YELLOW "?" RESET " or " YELLOW "-?" RESET "  to see help.\n");
		return 0;
	}

	create_project(&options);
	
	printf(BOLD_GREEN  "\n  apos: Project \"%s\" created!\n", options.project_name);
	printf(BOLD_MAGENTA"\n  Project Information:\n\n");
	printf(BOLD_CYAN"   Project Name:    "BOLD_YELLOW"\"%s\"\n", options.project_name);
	printf(BOLD_CYAN"   Git Repository:  "BOLD_YELLOW"%s\n", options.create_git ? "True":"False");
	printf(BOLD_CYAN"   Blink Template:  "BOLD_YELLOW"%s\n", options.blink_template ? "True":"False");
	printf(BOLD_CYAN"   Microcontroller: "BOLD_YELLOW"%s\n", options.device);
	printf(BOLD_CYAN"   CPU Frequency:   "BOLD_YELLOW"%s\n", options.f_cpu);
	printf(BOLD_CYAN"   Low Fuse:        "BOLD_YELLOW"%s\n", options.low_f);
	printf(BOLD_CYAN"   High Fuse:       "BOLD_YELLOW"%s\n", options.high_f);
	printf(BOLD_CYAN"   Extended Fuse:   "BOLD_YELLOW"%s\n", options.extended_f);
	printf(BOLD_CYAN"   Programmer:      "BOLD_YELLOW"%s\n\n", options.programmer);




	return 0;
}
